### PR TITLE
[ ACT ] ACTEDU-558: Implement login by username

### PR DESCRIPTION
**Pull request by OPersian**
**Description:** Enable login by username, preserving login by email.
See the docs on the use cases in the ticket.

**Youtrack:** https://youtrack.raccoongang.com/issue/ACTEDU-558

**Merge deadline:** 21 April 2020

### CODE REVIEW

File: common/static/common/js/utils/edx.utils.validate.js [hunk 1]
```diff
@@ -70,14 +75,43 @@
...
                         str: {
+
                             minlength: function($el) {
-                                var min = $el.attr('minlength') || 0;
+                                var min;
+                                var isEmailValue =  _fn.validate.email.regex.test($el.val());
+                                var isUsernameValue = _fn.validate.username.ascii_regex.test($el.val());
+
+                                if ($el.attr('type') === 'email') {
+                                    if (isEmailValue) {
```
Comment (OPersian) [inline common/static/common/js/utils/edx.utils.validate.js:85]: No sugar, easier to read!

File: common/static/common/js/utils/edx.utils.validate.js [hunk 2]
```diff
@@ -133,14 +179,18 @@
...
+                                isUsernameValue;
 
                             _.each(tests, function(value, key) {
                                 if (!value) {
                                     label = _fn.validate.getLabel($el.attr('id'));
                                     customMsg = $el.data('errormsg-' + key) || false;
                                     liveValidationMsg =
                                         $('#' + $el.attr('id') + '-validation-error-msg').text() || false;
+                                    isEmailValue =  _fn.validate.email.regex.test($el.val());
+                                    isUsernameValue = _fn.validate.username.ascii_regex.test($el.val());
```
Comment (GlugovGrGlib) [inline common/static/common/js/utils/edx.utils.validate.js:193]: Looks like `isUsernameValue` is not used in `getMessage` function. Please remove it if it is unused.

File: common/static/common/js/utils/edx.utils.validate.js [hunk 3]
```diff
@@ -102,6 +136,14 @@
                             return isBlank;
                         },
 
+                        // To support login by username
+                        username: {
+                            // In general case, username can only contain letters (A-Z, a-z), numerals (0-9), underscores (_), and hyphens (-)
+                            ascii_regex: new RegExp('^([A-Za-z0-9-_]+)$'),
+                            // If the `ENABLE_UNICODE_USERNAME` feature enabled, username can only contain letters, numerals, and @/./+/-/_ characters
+                            // unicode_regex: new RegExp('^([A-Za-z0-9_@.+-]+)$'),
```
Comment (GlugovGrGlib) [inline common/static/common/js/utils/edx.utils.validate.js:144]: Why unicode_regex is commented?

Comment (OPersian) [reply common/static/common/js/utils/edx.utils.validate.js:144]: Demoed the case with the unicode to Alex, but the ENABLE_UNICODE_USERNAME feature is disabled. I've actually never seen it enabled, but it's there.

File: common/static/common/js/utils/edx.utils.validate.js (hunk 2 above)
Comment (OPersian) [reply common/static/common/js/utils/edx.utils.validate.js]: Removed. @GlugovGrGlib please review again

**Comment by OPersian** [comment]
Duplicated  in https://gitlab.raccoongang.com/Duck-Team/projects/act-edu2050/edx-platform/-/merge_requests/3

**Pull request merged**